File: DentyEngine/src/Graphics/Color.cpp
#include "stdafx.h"

#include "Color.h"

namespace DentyEngine
{
	Color::Color() :
		color(WHITE)
	{
	}

	Color::Color(float r, float g, float b, float a) :
		color(r, g, b, a)
	{
	}

	Color::Color(const Vector4& color) :
		color(color)
	{
	}

	bool Color::IsApproxEqual(const Color& target) const noexcept
	{
		return color.IsApproxEqual(target.color);
	}

	Color Color::Lerp(const Color& a, const Color& b, float t)
	{
		Color result = { };
		result.color.x = std::lerp(a.color.x, b.color.x, t);
		result.color.y = std::lerp(a.color.y, b.color.y, t);
		result.color.z = std::lerp(a.color.z, b.color.z, t);
		result.color.w = std::lerp(a.color.w, b.color.w, t);

		return result;
	}
};

File: DentyEngine-ScriptCore/ScriptCore/Core/Color.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DentyEngine
{
    public class Color
    {
        public Color()
        {
            color = new Vector4(1.0f, 1.0f, 1.0f, 1.0f);
        }

        public Color(float r, float g, float b, float a)
        {
            color = new Vector4(r, g, b, a);
        }

        public Color(Vector4 color)
        {
            this.color = color;
        }

        public Color(Color color, bool transparent = false)
        {
            this.color = color.color;

            if (transparent)
            {
                this.color.w = 0.0f;
            }
        }

        //
        // Static variables.
        //
        public static Color BLACK
        {
            get
            {
                return new Color(0, 0, 0, 1);
            }
        }

        public static Color WHITE
        {
            get
            {
                return new Color(1, 1, 1, 1);
            }
        }

        public static Color RED
        {
            get
            {
                return new Color(1, 0, 0, 1);
            }
        }

        public static Color GREEN
        {
            get
            {
                return new Color(0, 1, 0, 1);
            }
        }

        public static Color BLUE
        {
            get
            {
                return new Color(0, 0, 1, 1);
            }
        }

        public static Color YELLOW
        {
            get
            {
                return new Color(1, 0.92f, 0.016f, 1);
            }
        }

        public static Color ORANGE
        {
            get
            {
                return new Color(0.952f, 0.59f, 0.0f, 1.0f);
            }
        }

        public static Color PINK
        {
            get
            {
                return new Color(1.0f, 0.75f, 0.79f, 1.0f);
            }
        }

        //
        // Member variables.
        //
        public Vector4 color;
    }
}